File: js/trees.js
var util = require('util');

var trees = (function() {
        var TYPES = {
            "number":0,
            "string":1,
            "boolean":2,
            "list":3,
            "dict":4,
            "null":5
        };

        var obj_to_json_type = function(k) {
            var t = typeof k;
            if (t == "number") {
                return "number";
            } else if (t == "string") {
                return "string";
            } else if (t == "boolean") {
                return "boolean";
            } else if (k.constructor == Array) {
                return "list";
            } else if (k.constructor == Object) {
                return "dict";
            } else if (k == null) {
                return "null";
            } else if (k.constructor == Node) {
                return "node";
            } else {
                throw "Unsupported data type: " + t;
            }
        };

        var type_for_enum = function(k) {
            for(var i in TYPES) {
                if (k == TYPES[i]) {
                    return TYPES[i];
                }
            }
            return null;
        };

        var Node = function(d) {
            if (d === undefined) {
                d = {};
            }

            this.name = null;
            this.type = null;
            this.attr = {};
            
            if (d['name'] != undefined) {
                this.name = d['name'];
            }
            if (d['type'] != undefined) {
                this.type = TYPES[d['type']];
            }
            if (d['type_id'] != undefined) {
                this.type = d['type'];
            }
            if (d['attr'] != undefined) {
                this.attr = d['attr'];
            }
            this.value = null;
            this.children = [];
        };

        Node.prototype.proto = function(d) {
            var value = d['value'] || null;
            var children = d['children'] || null;
            
            var p = new Node({"name":this.name, type_id:this.type, attr:this.attr});
            if (value !== null) {
                p.set_value(value);
            }
            if (children !== null) {
                p.children = children;
            }
            return p;
        };

        // Class method.
        Node.from_obj = function(obj) {
            var t = obj_to_json_type(obj);
            if (t == "number") {
                var n = new Node({"type":"number"});
                n.set_value(obj);
                return n;
            } else if (t == "string") {
                var n = new Node({"type":"string"});
                n.set_value(obj);
                return n;
            } else if (t == "boolean") {
                var n = new Node({"type":"boolean"});
                n.set_value(obj);
                return n;
            } else if (t == "list") {
                var n = new Node({"type":"list"});
                for(var index in obj) {
                    var value = obj[index];
                    var new_node = Node.from_obj(value);
                    n.children.push(new_node);
                }
                return n;
            } else if (t == "dict") {
                var n = new Node({"type":"dict"});
                for(var key in obj) {
                    var value = obj[key];
                    var new_node = new Node.from_obj(value);
                    new_node.attr['key'] = key;
                    n.children.push(new_node);
                }
                return n;
            } else if (t == "null") {
                return new Node({"type":"null"});
            }
        };
        
        Node.prototype.obj_repr = function() {
            if (this.type == TYPES["number"]) {
                return JSON.parse(this.value);
            } else if (this.type == TYPES["string"]) {
                return JSON.parse(this.value);
            } else if (this.type == TYPES["boolean"]) {
                return JSON.parse(this.value);
            } else if (this.type == TYPES["list"]) {
                var ret = [];
                for(var i in this.children) {
                    var node = this.children[i];
                    ret.push(node.obj_repr());
                }
                return ret;
            } else if (this.type == TYPES["dict"]) {
                var ret = {};
                for(var index in this.children) {
                    var node = this.children[index];
                    var key = node.attr["key"];
                    ret[key] = node.obj_repr();
                }
                return ret;
            } else if (this.type == TYPES["null"]) {
                return null;
            }
        };

        Node.prototype.set_value = function(obj) {
            var t = typeof obj;
            if (t == "number" || t == "string" || t == "boolean") {
                this.value = JSON.stringify(obj);
            } else if (t.constructor == Node) {
                this.value = obj.value;
            } else {
                this.children = Node.from_obj(obj);
            }
        };

        // Path stuff, this could change over time, this is nice to have now
        // but isn't as important for data representation (for modification it will be.
        
        Node.prototype.get_path = function(key) {
            var sp = key.split(".");
            if (sp.length == 1) {
                return this._get(sp[0]);
            } else {
                var obj = this._get(sp[0]);
                return obj.get_path(sp.slice(1).join("."));
            }
        };

        Node.prototype._get = function(key) {
            if (this.type == TYPES['dict']) {
                for(var i in this.children) {
                    var node = this.children[i];
                    var n_key = node.attr['key'];
                    if (key == n_key) {
                        return node;
                    }
                }
                throw "Key not found: " + key;
            } else if (Number(key) != NaN) {
                if (this.type == TYPES["string"]) {
                    return this.value[Number(key)];
                } else if (this.type == TYPES["list"]) {
                    return this.children[Number(key)];
                } else { 
                    throw "invalid path: " + key;
                }
            } else {
                throw "invalid path: " + key;
            }
        };

        Node.prototype.set_path = function(key, value) {
            var sp = key.split(".");
            if (sp.length == 1) {
                return this._set(sp[0], value);
            } else {
                var obj = this._get(sp[0]);
                return obj.set_path(sp.slice(1).join("."), value);
            }
        };

        Node.prototype._set = function(key, value) {
            if (this.type == TYPES['dict']) {
                for(var i in this.children) {
                    var obj = this.children[i];
                    if (key == obj.attr['key']) {
                        obj.set_value(value);
                        return;
                    }
                }
                // Looks like we didn't bail out during the for loop
                var n = new Node({'type':obj_to_json_type(value)});
                n.attr['key'] = key;
                n.set_value(value);
                this.children.push(n);
                return n;
            } else if (Number(key) != NaN) {
                var index = Number(key);
                if (this.type == TYPES['string']) {
                    this.value[index] = value;
                } else if (this.type == TYPES['list']) {
                    this.children[index] = Node.from_obj(value);
                } else {
                    throw "invalid path: " + key; 
                }
            } else {
                throw "invalid path: " + key; 
            }
        };

        Node.prototype.remove_path = function(key) {
            var sp = key.split(".");
            if (sp.length == 1) {
                return this._remove(sp[0])
            } else {
                var obj = this._get(sp[0]);
                return obj.remove_path(sp.slice(1).join("."));
            }
        };

        Node.prototype._remove = function(key) {
            if (this.type == TYPES['dict']) {
                var new_children = [];
                for(var i in this.children) {
                    var obj = this.children[i];
                    if (key != obj.attr['key']) {
                        new_children.push(obj);
                    }
                }
                this.children = new_children;
            } else if (Number(key) != NaN) {
                var index = Number(key);
                if (this.type == TYPES['string']) {
                    this.value = this.value.slice(0, index).concat(this.value.slice(index+1, this.value.length));
                } else if (this.type == TYPES['list']) {
                    this.children = this.children.slice(0, index).concat(this.children.slice(index+1, this.children.length));
                } else {
                    throw "invalid path: " + key;
                }
            } else {
                throw "invalid path: " + key;
            }
        };

        Node.prototype.toString = function() {
            return "<Node t:" + this.type + " attr: " + JSON.stringify(this.attr) + " value: " + JSON.stringify(this.value) + " children: " + JSON.stringify(this.children) + ">";
        };

        var exports = {"Node":Node, "TYPES":TYPES};
        // Lets see if we are running in the Node.js context or in a browser.
        if (typeof module !== 'undefined') {
            module.exports = exports;
        }
        return exports;
    })();

File: python/trees.py
import json

DEBUG = 1

TYPES = {
    'number':0,
    'string':1,
    'boolean':2,
    'list':3,
    'dict':4,
    'null':5
}

APPLY_TYPES = [
    TYPES['number'],
    TYPES['boolean'],
    TYPES['null']
]

#I've thought quite a bit about having more data types, things like
#sets, sorted_sets, packed_sets, 

def obj_to_json_type(k):
    if type(k) in (int, long, float, bool):
        return 'number'
    elif type(k) in (str, unicode):
        return 'string'
    elif type(k) in (bool,):
        return 'bool'
    elif type(k) in (list, ):
        return 'list'
    elif type(k) in (dict, ):
        return 'dict'
    elif k == None:
        return 'null'
    elif type(k) in (Node,):
        return 'node'
    else:
        raise Exception('unknown type %r' % str(type(k)))

def type_for_enum(k):
    for i in TYPES:
        if TYPES[i] == k:
            return i

class Node(object):
    def __init__(self, **kwargs):
        self.name = None
        self.type = None
        self.attr = {}

        if 'name' in kwargs:
            self.name = kwargs['name']
        if 'type' in kwargs:
            self.type = TYPES[kwargs['type']]
        elif 'type_id' in kwargs:
            self.type = kwargs['type_id']
        if 'attr' in kwargs:
            self.attr = kwargs['attr']

        self.value = None
        self.children = []

    def clone(self):
        return Node.from_obj(self.obj_repr())

    @classmethod
    def from_obj(cls, obj):
        t = obj_to_json_type(obj)
        if t == 'number':
            n = Node(type='number')
            n.set_value(obj)
            return n
        elif t == 'string':
            n = Node(type='string')
            n.set_value(obj)
            return n
        elif t == 'boolean':
            n = Node(type='boolean')
            n.set_value(obj)
            return n
        elif t == 'list':
            n = Node(type='list')
            for i in obj:
                n.children.append(Node.from_obj(i))
            return n
        elif t == 'dict':
            n = Node(type='dict')
            for k in obj:
                v = obj[k]
                nn = Node.from_obj(v)
                nn.attr['key'] = k
                n.children.append(nn)
            return n
        elif t == 'null':
            n = Node(type='null')
            return n
        elif t == 'node':
            return obj

    def __repr__(self):
        if DEBUG:
            return "<Node t:%s attr:%r value:%r children:%r>" % (
                type_for_enum(self.type), self.attr, self.value, self.children)
        else:
            return "<Node t:%s attr:%r value:%r children:%i>" % (
                type_for_enum(self.type), self.attr, self.value, len(self.children))

    def obj_repr(self):
        if self.type == TYPES['number']:
            if self.value == None:
                self.value = json.dumps(0)
            return json.loads(self.value)
        elif self.type == TYPES['string']:
            if self.value == None:
                self.value = json.dumps('')
            return json.loads(self.value)
        elif self.type == TYPES['boolean']:
            if self.value == None:
                self.value = json.dumps(False)
            return json.loads(self.value)
        elif self.type == TYPES['list']:
            return [i.obj_repr() for i in self.children]
        elif self.type == TYPES['dict']:
            return dict([(i.attr['key'], i.obj_repr()) for i in self.children])
        elif self.type == TYPES['null']:
            return None
        else:
            raise Exception("%i is not defined in the TYPES table" % self.type)


    @classmethod
    def default_for_type(self, t):
        if t == TYPES['number']:
            return 0
        elif t == TYPES['string']:
            return ''
        elif t == TYPES['boolean']:
            return False
        elif t == TYPES['list']:
            return []
        elif t == TYPES['dict']:
            return dict()
        elif t == TYPES['null']:
            return None
        else:
            raise Exception("%i is not defined in the TYPES table" % t)

    def set_value(self, obj):
        if type(obj) in (int, long, float, bool, str, unicode):
            self.value = json.dumps(obj)
        elif type(obj) in (Node,):
            if obj.value != None:
                self.value = obj.value
            if obj.children != None:
                self.children = obj.children
            else:
                raise Exception("Bad set value")
        else:
            self.value = None

    ## Nice things to have for testing.
    def test_path(self, key):
        if not key:
            return True
        if len(key) == 1:
            return self._test(key[0])
        else:
            obj = self._test(key[0])
            if type(obj) == bool:
                return obj
            return obj.test_path(key[1:])

    def _test(self, key):
        if self.type == TYPES['dict']:
            for i in self.children:
                if key == i.attr['key']:
                    return True
            return False
        elif type(key) in (int, long):
            if self.type == TYPES['string']:
                return True
            elif self.type == TYPES['list']:
                return True
            else:
                return False
        else:
            return False

    def get_path(self, key):
        if not key:
            return self
        if len(key) == 1:
            return self._get(key[0])
        else:
            obj = self._get(key[0])
            return obj.get_path(key[1:])

    def _get(self, key):
        if self.type == TYPES['dict']:
            for i in self.children:
                if key == i.attr['key']:
                    return i
            raise Exception('key not found %s' % key)
        elif type(key) in (int, long):
            if self.type == TYPES['string']:
                return self.value[int(key)]
            elif self.type == TYPES['list']:
                return self.children[int(key)]
            else:
                raise Exception('Invalid path: cannot index type %s' % self.type)
        else:
            raise Exception('invalid path')

    def set_path(self, key, val):
        if not key:
            return self.set_value(val)
        if len(key) == 1:
            return self._set(key[0], val)
        else:
            obj = self._get(key[0])
            return obj.set_path(key[1:], val)

    def _set(self, key, value):
        if self.type == TYPES['dict']:
            for i in self.children:
                if key == i.attr['key']:
                    i.set_value(value)
                    return

            #looks like we didn't find the key
            if type(value) != Node:
                n = Node(type=obj_to_json_type(value))
            else:
                n = value.clone()
            n.attr['key'] = key
            n.set_value(value)
            self.children.append(n)
            return n
        elif type(key) in (int, long):
            if self.type == TYPES['string']:
                self.value[int(key)] = value
            elif self.type == TYPES['list']:
                if int(key) > len(self.children):
                    self.children.insert(int(key), Node.from_obj(value))
                else:
                    self.children[int(key)] = Node.from_obj(value)
            else:
                raise Exception('Invalid path: cannot index type %s' % self.type)
        else:
            raise Exception('invalid path')

    def remove_path(self, key):
        if len(key) == 1:
            return self._remove(key[0])
        else:
            obj = self._get(key[0])
            return obj.remove_path(key[1:])

    def _remove(self, key):
        if self.type == TYPES['dict']:
            new_children = []
            for i in self.children:
                if key != i.attr['key']:
                    new_children.append(i);
            self.children = new_children;
        elif type(key) in (int, long):
            if self.type == TYPES['string']:
                raise Exception('not supported')
            elif self.type == TYPES['list']:
                ikey = int(key)
                if ikey == 0:
                    self.children = self.children[ikey+1:]
                else:
                    self.children = self.children[:ikey] + self.children[ikey+1:]
            else:
                raise Exception('Invalid path: cannot index type %s' % self.type)
        else:
            raise Exception('invalid path')
